### PR TITLE
Assignment6

### PR DESCRIPTION
**Pull request by gknutson1**

### CODE REVIEW

File: Calculate/Program.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    public Program()
+    {
+        WriteLine = Console.WriteLine;
+        ReadLine = Console.ReadLine!;
+    }
+
+    public WriteDelegator WriteLine { get; }
+    public ReadDelegator ReadLine { get; }
+
```
Comment (alowe4) [inline Calculate/Program.cs:23]: both WriteLine and ReadLine need to be init only setters

File: Tests/Tests.csproj [hunk 2]
```diff
@@ -0,0 +1,28 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>net7.0</TargetFramework>
+        <ImplicitUsings>enable</ImplicitUsings>
+        <LangVersion>11</LangVersion>
+        <EnableNETAnalyzers>true</EnableNETAnalyzers>
```
Comment (alowe4) [inline Tests/Tests.csproj:7]: need directory.build.props file

File: Calculate/Calculate.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
+﻿namespace Calculate;
+
+public class Calculate
+{
```
Comment (alowe4) [inline Calculate/Calculate.cs:4]: class not named correctly, should be named Calculator

File: Calculate/Calculate.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        new Dictionary<char, OperationDelegator>
+        {
+            { '+', Add },
+            { '-', Subtract },
+            { '*', Multiple },
+            { '/', Divide }
+        };
+
+    private Program.WriteDelegator WriteLine { get; }
+    private Program.ReadDelegator ReadLine { get; }
```
Comment (alowe4) [inline Calculate/Calculate.cs:21]: i don't understand why this is here, why does the calculator need to be able to read and write lines? shouldn't it only need to do a calculation based on what it's given?

**Review by alowe4 — COMMENTED** [review]
Instructions
    Create a Console project called "Calculate.". ✔
    Define a Program Class
        Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ❌ [not init only]
        Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
        Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
    Define a Calculator class ❌ [named wrong]
        Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
        Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:✔
            is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
                Uses char for the key corresponding to the operators +, -, *, and /. ✔
                Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
        Implement a TryCalculate method following "TryParse" pattern ✔
            Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
            If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
            Use string.Split(), pattern matching, logical and operators to parse the string in their entirety.✔
            Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
    Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
    Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔

Extra Credit
Do one of the following two options (or both if you want extra, extra credit) :)
    Refactor the redirect portion of the Program class into 'ProgramBase`❌
    Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects❌
    Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0)❌

 Fundamentals
    Place all shared project properties into a Directory.Build.Props file.❌
    Place all shared project items into a Directory.Build.targets file.❌
    nullable reference types is enabled ✔
    Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
    Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
    and enabled .NET analyzers for both projects ✔
    For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
    All of the above should be unit tested ✔
    Choose simplicity over complexity ✔

File: Calculate/Program.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
+﻿namespace Calculate;
+
+public class Program
+{
+    public delegate string ReadDelegator();
+
+    public delegate void WriteDelegator(string message);
+
+    public Program(WriteDelegator writeLine, ReadDelegator readLine)
```
Comment (mmwoodfo) [inline Calculate/Program.cs:9]: Good job adding a constructor for setting these, but these should be func and action types

File: Calculate/Calculate.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    private static int Add(int a, int b) => a + b;
+    private static int Subtract(int a, int b) => a - b;
+    private static int Multiple(int a, int b) => a * b;
+    private static int Divide(int a, int b) => a / b;
+
+
+    public bool TryCalculate()
+    {
+        string calculation = ReadLine();
+        string[] parts = calculation.Split(' ', StringSplitOptions.RemoveEmptyEntries);
```
Comment (mmwoodfo) [inline Calculate/Calculate.cs:32]: You should be using the Split method along side pattern matching. e.g.
`if (expression?.Split(" ") is [string operator1, [char mathmaticalOperator], string operand2])`

File: Tests/UnitTest1.cs [hunk 7]
```diff
@@ -0,0 +1,124 @@
+using Calculate;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Tests;
+
+[TestClass]
```
Comment (mmwoodfo) [inline Tests/UnitTest1.cs:6]: You could use some data driven tests here to cover more cases

**Pull request closed without merging**